File: lox/scanner/scanner.cpp
#include <lox/errors.hpp>
#include <lox/scanner/scanner.hpp>

#include <fmt/core.h>

namespace lox {

static auto get_token_type(std::string_view ident) -> token_type {
  if (ident == "and") return token_type::AND;
  if (ident == "class") return token_type::CLASS;
  if (ident == "else") return token_type::ELSE;
  if (ident == "false") return token_type::FALSE;
  if (ident == "for") return token_type::FOR;
  if (ident == "fun") return token_type::FUN;
  if (ident == "if") return token_type::IF;
  if (ident == "nil") return token_type::NIL;
  if (ident == "or") return token_type::OR;
  if (ident == "print") return token_type::PRINT;
  if (ident == "return") return token_type::RETURN;
  if (ident == "super") return token_type::SUPER;
  if (ident == "this") return token_type::THIS;
  if (ident == "true") return token_type::TRUE;
  if (ident == "var") return token_type::VAR;
  if (ident == "while") return token_type::WHILE;
  if (ident == "break") return token_type::BREAK;

  return token_type::IDENTIFIER;
}

// clang-format off
static auto is_digit(char c) -> bool { return c >= '0' && c <= '9'; }
static auto is_alpha(char c) -> bool { return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'; }
static auto is_alphanumeric(char c) -> bool { return is_alpha(c) || is_digit(c); }
// clang-format on

auto scanner::scan() -> std::vector<token> {
  while (!done()) {
    start_ = curr_;
    scan_next();
  }

  tokens_.push_back(token{token_type::EOF, "", line_});
  return tokens_;
}

void scanner::scan_next() {
  char c = next();
  using enum lox::token_type; // C++20 addition!
  switch (c) {
  // clang-format off
  case '(': add_token(LEFT_PAREN); break;
  case ')': add_token(RIGHT_PAREN); break;
  case '{': add_token(LEFT_BRACE); break;
  case '}': add_token(RIGHT_BRACE); break;
  case ',': add_token(COMMA); break;
  case '.': add_token(DOT); break;
  case '-': add_token(MINUS); break;
  case '+': add_token(PLUS); break;
  case ';': add_token(SEMICOLON); break;
  case '*': add_token(STAR); break;
  case '?': add_token(QUESTION); break;
  case ':': add_token(COLON); break;

  case '"': string(); break;
    // clang-format on

  case '!':
    add_token(match('=') ? BANG_EQUAL : BANG);
    break;
  case '=':
    add_token(match('=') ? EQUAL_EQUAL : EQUAL);
    break;
  case '<':
    add_token(match('=') ? LESS_EQUAL : LESS);
    break;
  case '>':
    add_token(match('=') ? GREATER_EQUAL : GREATER);
    break;
  case '/':
    if (match('/')) {
      // A comment goes until the end of the line
      while (peek() != '\n' && !done()) next();
    } else if (match('*')) {
      // Find closing */
      block_comment();
    } else {
      add_token(SLASH);
    }
    break;

  case ' ':
  case '\r':
  case '\t':
    // Ignore whitespace
    break;
  case '\n':
    ++line_;
    break;

  case 'o':
    if (match('r')) { add_token(OR); }
    break;

  default:
    if (is_digit(c)) {
      number();
    } else if (is_alpha(c)) {
      identifier();
    } else {
      error::report(line_, fmt::format("unexpected character: {}", c));
    }
    break;
  }
}

void scanner::string() {
  while (peek() != '"' && !done()) {
    if (peek() == '\n') ++line_;
    next();
  }

  if (done()) {
    error::report(
        line_, fmt::format("unterminated string: {}", substr(start_, curr_)));
    return;
  }

  // The closing "
  next();

  // Trim the surrounding quotes
  add_token(token_type::STRING, substr(start_ + 1, curr_ - 1));
}

void scanner::number() {
  while (is_digit(peek())) next();

  // Look for a fractional part
  if (peek() == '.' && is_digit(peek_next())) {
    // Consume the "."
    next();

    while (is_digit(peek())) next();
  }

  double num = std::stod(substr(start_, curr_));
  add_token(token_type::NUMBER, num);
}

void scanner::identifier() {
  while (is_alphanumeric(peek())) next();

  token_type type = get_token_type(substr(start_, curr_));
  add_token(type);
}

void scanner::block_comment() {
  // A naive implementation searches for the closing */ but that may overlap
  // a nested comment!
  // Idea: keep track of the levels of nesting.
  for (int depth = 1; depth > 0;) {
    if (done()) {
      error::report(line_, "unterminated block comment");
      return;
    }

    if (match('/') && peek() == '*') {
      next();
      ++depth;
      continue;
    }

    if (match('*') && peek() == '/') {
      next();
      --depth;
      continue;
    }

    if (match('\n')) {
      line_++;
      continue;
    }

    next();
  }
}

} // namespace lox

File: glox/internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"unicode"

	"github.com/tzcl/lox/glox/internal/token"
)

type ScannerError struct {
	line    int
	message string
}

func (e *ScannerError) Error() string {
	return fmt.Sprintf("[line %d]: Error: %s", e.line, e.message)
}

type Scanner struct {
	// Storing the source as a slice of runes allows us to handle Unicode
	// characters
	source []rune

	start int
	curr  int
	line  int

	tokens []token.Token
}

func New(source string) *Scanner {
	return &Scanner{source: []rune(source), line: 1}
}

func (s *Scanner) Scan() ([]token.Token, error) {
	for !s.done() {
		if err := s.scanToken(); err != nil {
			return nil, err
		}
	}

	s.tokens = append(s.tokens, token.Token{Type: token.EOF, Line: s.line})
	return s.tokens, nil
}

func (s *Scanner) scanToken() error {
	s.start = s.curr

	switch r := s.next(); r {
	// Single-character tokens
	case '(':
		s.addToken(token.LeftParen)
	case ')':
		s.addToken(token.RightParen)
	case '{':
		s.addToken(token.LeftBrace)
	case '}':
		s.addToken(token.RightBrace)
	case ',':
		s.addToken(token.Comma)
	case '.':
		s.addToken(token.Dot)
	case '-':
		s.addToken(token.Minus)
	case '+':
		s.addToken(token.Plus)
	case ';':
		s.addToken(token.Semicolon)
	case '?':
		s.addToken(token.QuestionMark)
	case ':':
		s.addToken(token.Colon)

	// Double-character tokens
	case '!':
		ttype := token.Bang
		if s.match('=') {
			ttype = token.BangEqual
		}
		s.addToken(ttype)
	case '=':
		ttype := token.Equal
		if s.match('=') {
			ttype = token.EqualEqual
		}
		s.addToken(ttype)
	case '<':
		ttype := token.Less
		if s.match('=') {
			ttype = token.LessEqual
		}
		s.addToken(ttype)
	case '>':
		ttype := token.Greater
		if s.match('=') {
			ttype = token.GreaterEqual
		}
		s.addToken(ttype)
	case '/':
		switch n := s.peek(); n {
		case '/':
			s.next() // Consume the '/'
			for s.peek() != '\n' && !s.done() {
				s.next()
			}
		case '*':
			s.next() // Consume the '*'
			if err := s.blockComment(); err != nil {
				return err
			}
		default:
			s.addToken(token.Slash)
		}
	case '*':
		if s.peek() == '/' {
			return s.error("found block comment without matching /*")
		}
		s.addToken(token.Star)

	// Ignore whitespace
	case ' ', '\r', '\t':
	case '\n':
		s.line++

	// Parse string
	case '"':
		if err := s.string(); err != nil {
			return err
		}

	// Parse more complex tokens
	default:
		switch {
		case unicode.IsDigit(r):
			s.number()
		case isLetter(r):
			if err := s.identifier(); err != nil {
				return err
			}
		default:
			return s.error("unexpected character")
		}
	}

	return nil
}

func (s *Scanner) string() error {
	// Find end of string
	for s.peek() != '"' && !s.done() {
		if s.peek() == '\n' {
			s.line++
		}
		s.next()
	}

	if s.done() {
		str := string(s.source[s.start:s.curr])
		return s.error("unterminated string: " + str)
	}

	// The closing "
	s.next()

	// Trim the surrounding quotes
	str := string(s.source[s.start+1 : s.curr-1])
	s.addTokenWithLexeme(token.String, str)

	return nil
}

func (s *Scanner) number() {
	// Find end of number
	for unicode.IsDigit(s.peek()) {
		s.next()
	}

	// Look for fractional part
	if s.peek() == '.' && unicode.IsDigit(s.peekNext()) {
		// Consume the "."
		s.next()

		// Find end of fractional part
		for unicode.IsDigit(s.peek()) {
			s.next()
		}
	}

	number := string(s.source[s.start:s.curr])
	s.addTokenWithLexeme(token.Number, number)
}

func (s *Scanner) identifier() error {
	// Find end of identifier
	for r := s.peek(); isLetter(r) || unicode.IsDigit(r); r = s.peek() {
		s.next()
	}

	if s.match('"') {
		ident := string(s.source[s.start:s.curr])
		return s.error(`string missing opening quote: ` + ident)
	}

	ident := string(s.source[s.start:s.curr])
	ttype := token.LookupKeyword(ident)
	s.addTokenWithLexeme(ttype, ident)

	return nil
}

func (s *Scanner) blockComment() error {
	var depth int

	for !s.done() {
		switch {
		case s.match('/') && s.match('*'):
			depth++
		case s.match('*') && s.match('/'):
			depth--
			if depth < 0 {
				return nil
			}
		case s.match('\n'):
			s.line++
		default:
			s.next()
		}
	}

	return s.error("unterminated block comment")
}

func (s *Scanner) done() bool {
	return s.curr >= len(s.source)
}

func (s *Scanner) match(r rune) bool {
	if s.done() {
		return false
	}
	if s.source[s.curr] != r {
		return false
	}

	s.curr++
	return true
}

func (s *Scanner) next() rune {
	r := s.source[s.curr]
	s.curr++
	return r
}

func (s *Scanner) peek() rune {
	if s.done() {
		// rune(0) == '\000' == '\x00' == '\u0000'
		return rune(0)
	}

	return s.source[s.curr]
}

func (s *Scanner) peekNext() rune {
	if s.curr+1 >= len(s.source) {
		return rune(0)
	}

	return s.source[s.curr+1]
}

func (s *Scanner) addToken(ttype token.Type) {
	s.addTokenWithLexeme(ttype, string(s.source[s.start:s.curr]))
}

func (s *Scanner) addTokenWithLexeme(ttype token.Type, lexeme string) {
	token := token.Token{Type: ttype, Lexeme: lexeme, Line: s.line}
	s.tokens = append(s.tokens, token)

}

func (s *Scanner) error(message string) error {
	return &ScannerError{line: s.line, message: message}
}

// A letter can be a unicode letter or "_"
func isLetter(r rune) bool {
	return unicode.IsLetter(r) || r == '_'
}
